src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +68,44 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+/** @brief Convert an integer value into printable base-N representation.
+ * @param[in] value Machine integer for conversion
+ * @param[out] str Pointer to caller-managed buffer write the string into
+ * @param[in] base Representation base between 2 and 36
+ * @returns NULL on error (unsupported base), or pointer into str,
+ *          advanced past the written part (for chaining; points to inserted NUL-terminator)
+ */
+static char *utoa_upper(uint32_t value, char *const str, const int base)
+{
+	static const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+	static char buf[33] = {0};
+	size_t total = 0;
+
+	/* Check base is supported. */
+	if (base < 2 || base > 36) {
+		str[0] = '\0';
+		return NULL;
+	}
+
+	/* Push up to 32 symbols (base-2 aka binary, worst case) */
+	for (size_t offset = 0; offset < 32; ++offset) {
+		const div_t result = div(value, base);
+		const uint32_t remainder = result.rem;
+		buf[offset] = digits[remainder];
+		if (result.quot == 0) {
+			total = offset;
```

Comment:
If this is `offset + 1U`, then the nul terminations become `buf[total]` and `str[total]` and the loop below just `i < total`, - this should be a good optimisation.

Reply:
Traded three instances of +1 (and `s/<=/</`) with a slightly more convoluted indexing expression in second loop. That's what I tried to avoid. It was mentally easier to me to perceive these +1's as explicitly dealing with the 0-termination.

Reply:
Ah, we hadn't spotted that `total - i` part.. 🤔 that does make it more difficult. This does make another good argument for dispensing with the nul termination in the first buffer though.. as much as we understand why you don't want to do that re debugging.

Reply:
Re dropping nul terminator: I look on this as a string function, it quacks like a string function (inverse of atoi) -- you give it an integer/countable and a buffer, it produces a printable array of symbols aka a message string. So I expect someone to potentially try to misuse it elsewhere, and extra nuls might help protect someone's feet.
Makes me wonder why Dfu St Extensions string looks like a human-readable description of target flash (unrelated, we're dealing with USB serialno).

Reply:
To be clear, we don't mean dropping the final nul termination, we mean dropping the intermediate buffer nul termination - `char buf[32]`, line 102 goes, that kind of thing.

Reply:
Okay, that is minor, I could drop that line, sacrificing some internals' debuggability on multiple calls, but DFU doesn't make any (not that I know of).

Reply:
Since that intermediate buffer is private / insignificant, I pulled a sneaky and partially reverted the logic around `total = offset` to mostly avoid any +1/-1. Tested live, the last preincrement is required.